common/concurrency/ConcurrentTask.h

```diff
@@ -0,0 +1,108 @@
+    ConstExprStr metricName;
+    ConstExprStr workerMetricName;
+    const size_t size;
+    spdlog::logger &tracer;
+    WorkerPool &workers;
+
+public:
+    ConcurrentTask(ConstExprStr metricName, ConstExprStr workerMetricName, size_t size, spdlog::logger &tracer,
+                   WorkerPool &workers)
+        : inputq(std::make_shared<ConcurrentBoundedQueue<I>>(size)), metricName(metricName),
+          workerMetricName(workerMetricName), size(size), tracer(tracer), workers(workers) {}
+
+    void enqueue(I &&input) {
+        inputq->push(std::move(input), 1);
+    }
+
+    void run(std::function<void(I &&, WorkerState &)> processTask, std::function<void(WorkerState &&)> combineOutput) {
+        ENFORCE_NO_TIMER(inputq->hasAllInput(), "Input queue is not filled; did you specify the wrong queue size?");
+        Timer timeit(tracer, metricName);
+
+        auto outputq = std::make_shared<BlockingBoundedQueue<WorkerState>>(size);
+        std::shared_ptr<BlockingBoundedQueue<CounterState>> counters;
+        if (CollectCounters) {
+            // If workers is empty, the main thread will run the job.
+            counters = std::make_shared<BlockingBoundedQueue<CounterState>>(std::max(workers.size(), 1));
+        }
+
+        // Note: Cannot be stack allocated; see docs for absl::Barrier. The +1 is for the control thread.
+        auto workerBarrier = new absl::Barrier(workers.size() + 1);
+
+        // Note: We must copy `barrier` pointer into the lambda, as `delete barrier` races with the coordinator thread
+        // destroying the stack frame.
+        workers.multiplexJob(metricName.str, [&tracer = this->tracer, &workerMetricName = this->workerMetricName,
+                                              inputq = this->inputq, outputq, processTask, counters,
+                                              &workers = this->workers, barrier = workerBarrier]() {
+            {
+                Timer timeit(tracer, workerMetricName);
+                I input;
+                size_t count = 0;
+                WorkerState state;
+                for (auto result = inputq->try_pop(input); !result.done(); result = inputq->try_pop(input)) {
+                    if (result.gotItem()) {
+                        count++;
+                        processTask(std::move(input), state);
+                    }
+                }
+                if (count > 0) {
+                    outputq->push(std::move(state), count);
+                }
+
+                if (CollectCounters) {
+                    counters->push(getAndClearThreadCounters(), 1);
+                }
+            }
+
+            // Prevent deadlock with an empty workerpool, where all logic runs on one thread.
+            // After this LOC, it is no longer safe to use `this` or refer to any variables in the `run` stack
+            // frame.
+            if (workers.size() > 0 && barrier->Block()) {
+                delete barrier;
```

Comment:
can we do without raw pointer delete ops?

Reply:
can we make unique_ptr instead?

Reply:
I'm not sure if I can. This is the pattern the absl docs want you to use. I fear that unique_ptr delete may race with worker threads waking up from block.

Reply:
I might be able to hide this in an RAII class.